[vulkan] Fix warnings: static_cast, remove unused

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42195 [vulkan] Fix warnings: static_cast, remove unused**

Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

IvanKobzarev committed:
[vulkan] Fix warnings: static_cast, remove unused

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/Vulkan.cpp:
```diff
@@ -989,7 +989,8 @@ void copy_buffer_to_image(const VBuffer& buffer, VImage& image) {
   };
-  ConstBlock constBlock{image.w(), image.h()};
+  ConstBlock constBlock{static_cast<int32_t>(image.w()),
+                        static_cast<int32_t>(image.h())};
```
I think safe_downcast is a better choice here but that might requires some refactoring to move it out of Pool.h.

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanCommon.h:
```diff
@@ -54,6 +54,21 @@ struct ContextConv2D final {
 
+namespace detail {
+template <typename To, typename From>
+inline constexpr To safe_downcast(From v) {
```
I'm not sure this can be  constexpr anymore considering that some implementations of TORCH_CHECK might throw so if you run into any compilation issues, that's a potential reason.

AshkanAliabadi replied:
Also please mark `v` as const.

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanCommon.h:
```diff
@@ -54,6 +54,21 @@ struct ContextConv2D final {
+  constexpr Type max{static_cast<Type>(std::numeric_limits<To>::max())};
+  TORCH_CHECK(min <= v && v <= max, "Cast failed: out of range");
+  return static_cast<To>(v);
+}
```
This function should probably be in c10 since it is general purpose.  You can talk with David, Dima, and c10 maintainers if you are OK with moving it there.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -473,8 +490,10 @@ ImageSizes conv2d_prepack_weights_image_sizes(
     int64_t KW) {
-  return {{ALIGN_UP4(C), UP_DIV(OC, 4), KH * KW},
-          {ALIGN_UP4(C), UP_DIV(OC, 4), KH * KW}};
+  const uint32_t Cup4 = ALIGN_UP4(static_cast<uint32_t>(C));
```
I should have mentioned this earlier, but inline functions (and functions in general) are superior to macros.  A change you can consider in a later PR.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -627,7 +653,7 @@ void conv2d(
   constBuffer.bind(descriptorSet, 4);
 
-  WorkGroupSize workGroupSize{1, 1, params.OC_4};
+  WorkGroupSize workGroupSize{1, 1, static_cast<uint32_t>(params.OC_4)};
```
safe_downcast?

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -473,8 +490,10 @@ ImageSizes conv2d_prepack_weights_image_sizes(
-  return {{ALIGN_UP4(C), UP_DIV(OC, 4), KH * KW},
-          {ALIGN_UP4(C), UP_DIV(OC, 4), KH * KW}};
+  const uint32_t Cup4 = ALIGN_UP4(static_cast<uint32_t>(C));
+  const uint32_t OC_4 = UP_DIV(static_cast<uint32_t>(OC), 4);
```
safe_downcast?

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -473,8 +490,10 @@ ImageSizes conv2d_prepack_weights_image_sizes(
-          {ALIGN_UP4(C), UP_DIV(OC, 4), KH * KW}};
+  const uint32_t Cup4 = ALIGN_UP4(static_cast<uint32_t>(C));
+  const uint32_t OC_4 = UP_DIV(static_cast<uint32_t>(OC), 4);
+  const uint32_t Z = static_cast<uint32_t>(KH) * static_cast<uint32_t>(KW);
```
ditto

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanCommon.h:
```diff
@@ -54,6 +54,21 @@ struct ContextConv2D final {
+namespace detail {
+template <typename To, typename From>
+inline constexpr To safe_downcast(From v) {
+  if (std::is_signed<From>::value && std::is_unsigned<To>::value) {
```
Also this is more of an advanced topic (just a tip - you don't need to make the change), but since this condition is completely a compile-time condition, you could either use SFINAE in pre C++-17 or `if constexpr` in C++-17 and onward to isolate it.  That would also prevent against compilation errors in case this if block contained a block of code that would only make sense (i.e. was only compilable) if this condition held true but not in general.  This is not the case here.

IvanKobzarev replied:
I have updated PR with enable_if construction :

```
template <typename To, typename From>
inline constexpr To safe_downcast_internal(From v) {
  typedef std::common_type_t<From, To> Type;
  constexpr Type min{static_cast<Type>(std::numeric_limits<To>::lowest())};
  const Type value{static_cast<Type>(v)};
  constexpr Type max{static_cast<Type>(std::numeric_limits<To>::max())};
  TORCH_CHECK(min <= v && v <= max, "Cast failed: out of range");
  return static_cast<To>(v);
}

template <typename To, typename From>
inline constexpr bool IsSignedToUnsigned() {
  return std::is_signed<From>::value && std::is_unsigned<To>::value;
}

template <
    typename To,
    typename From,
    std::enable_if_t<IsSignedToUnsigned<To, From>(), bool> = true>
inline constexpr To safe_downcast(From v) {
  TORCH_CHECK(v >= From{}, "Cast failed: negative signed to unsigned");
  return safe_downcast_internal<To, From>(v);
}

template <
    typename To,
    typename From,
    std::enable_if_t<!IsSignedToUnsigned<To, From>(), bool> = true>
inline constexpr To safe_downcast(From v) {
  return safe_downcast_internal<To, From>(v);
}
```

AshkanAliabadi commented:
By the way, I also suggest calling this numeric_cast.  Downcast implies some sort of inheritance hierarchy.

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Fix warnings: static_cast, remove unused"


Differential Revision: [D22803035](https://our.internmc.facebook.com/intern/diff/D22803035)

Fix warnings for internal BUCK build that has stricter compilation flags (in comparison with oss build).
 - static_cast<int32_t> where int64_t  sizes, strides are used in Vulkan shaders as int32_t
 - unused variables in `vulkan_test.cpp`

[ghstack-poisoned]
